xcm precompile transfers (#174)
moondance-labs/tanssi-docs: builders/toolkit/ethereum-api/precompiles/cross-chain-transfers.md

---
title: Native Cross-Chain Token Transfers
description: Learn how to use the XCM interface precompile to transfer tokens from any Tanssi EVM appchain, leveraging their inherent native cross-chain capabilities.
---

# Native Cross-Chain Token Transfers

## Introduction {: #introduction }

As presented in the [Native Cross-Chain Communication](/learn/framework/xcm/){target=\_blank} article from the Learn section, Tanssi appchains benefit from an inherent capability to communicate and interoperate with any other appchain in the ecosystem. This native cross-chain communication allows safe and fast token transfers leveraging the Cross-Consensus Message format (XCM for short), which facilitates communication between different consensus systems.

The communication protocol enabling token transfers is built on [Substrate](/learn/framework/overview/#substrate-framework){target=\_blank} and runs on a lower level than the EVM, making it harder for EVM developers to access.

> Nevertheless, EVM Appchains have an XCM precompile that fills the gap between execution layers, exposing a smart contract interface that abstracts away the underlying complexities, making the execution of cross-chain token transfers as easy as any other smart contract call.

Reviewer: ```suggestion
Nevertheless, EVM appchains have an XCM precompile that fills the gap between execution layers, exposing a smart contract interface that abstracts away the underlying complexities, making the execution of cross-chain token transfers as easy as any other smart contract call.
```

> This guide will show you how to interact with the [XCM interface](https://github.com/moondance-labs/tanssi/blob/master/test/contracts/solidity/XcmInterface.sol){target=\_blank} precompile to execute cross-chain token transfers through the Ethereum API.

Reviewer: ```suggestion
This guide will show you how to interact with the [XCM Interface](https://github.com/moondance-labs/tanssi/blob/master/test/contracts/solidity/XcmInterface.sol){target=\_blank} precompile to execute cross-chain token transfers through the Ethereum API.
```

Reviewer: The reason why Interface should be capitalized here is because we're referring to the contract name

> The XCM precompile is located at the following address:

Author: Oh, there was a change I didn't commit 😅

```text

> {{networks.dancebox.precompiles.xcmInterface }}

Reviewer: The variable was updated in the `variables.yml` file, but also needs to be updated here where you use it

```

--8<-- 'text/builders/toolkit/ethereum-api/precompiles/security-note.md'

## The XCM Solidity Interface {: #the-xcm-solidity-interface }

The [`XCMInterface.sol`](https://github.com/moondance-labs/tanssi/blob/master/test/contracts/solidity/XcmInterface.sol){target=\_blank} interface on Tanssi EVM appchains is a Solidity interface that allows developers to interact with the precompile's functions.

??? code "XCMInterface.sol"

    ```solidity
    --8<-- 'code/builders/toolkit/ethereum-api/precompiles/xcm-interface/XcmInterface.sol'
    ```

The interface includes the necessary data structures along with the following functions:

> ???+ function "**transferAssetsToPara20**(_paraId, beneficiary, assets, feeAssetItem, weight_) — sends assets to another EVM-compatible appchain using the underlying `transfer_assets()` transaction included in the XCM pallet module"

Author: [This PR](https://github.com/papermoonio/tanssi-mkdocs/pull/37) adding the styles was merged only four days ago, maybe you need to pull the latest version of the mkdocs repo?

> ???+ function "**transferAssetsToPara20**(_paraId, beneficiary, assets, feeAssetItem, weight_) — sends assets to another EVM-compatible appchain using the underlying `transfer_assets()` transaction included in the XCM pallet module"

Reviewer 2: You are correct! Pulled the new one and now it looks great.

        - `feeAssetItem` - 0
        - `weight` - [9223372036854775807, 9223372036854775807]
    
## Interact with the Solidity Interface {: #interact-with-the-solidity-interface }

### Checking Prerequisites {: #checking-prerequisites }

To follow along with this tutorial, you will need to have your wallet configured to work with your EVM appchain and an account funded with native tokens. You can add your EVM appchain to MetaMask with one click on the [Tanssi dApp](https://apps.tanssi.network){target=\_blank}. Or, you can [configure MetaMask for Tanssi with the demo EVM appchain](/builders/toolkit/ethereum-api/wallets/metamask/){target=\_blank}.

!!! note
    It is necessary to have previously established communication channels with the destination chain before using this precompile's functionality. To do so, refer to the [Manage Cross-Chain Communication Channels](/builders/manage/dapp/xcm-channels/){target=\_blank} guide.
    Also, if the token being transferred is native to your appchain, the destination chain must have registered the foreign asset.

### Remix Set Up {: #remix-set-up }

> You can interact with the XCM interface precompile using [Remix](https://remix.ethereum.org){target=\_blank}. To add the precompile to Remix, you will need to:

Reviewer: ```suggestion
You can interact with the XCM Interface precompile using [Remix](https://remix.ethereum.org){target=\_blank}. To add the precompile to Remix, you will need to:
```

> 1. Get a copy of [`XcmInterface.sol`](https://github.com/moondance-labs/tanssi/blob/master/test/contracts/solidity/XcmInterface.sol){target=\_blank}

Reviewer 3: ```suggestion
1. Get a copy of [`XCMInterface.sol`](https://github.com/moondance-labs/tanssi/blob/master/test/contracts/solidity/XcmInterface.sol){target=\_blank}
```

2. Paste the file contents into a Remix file named `XcmInterface.sol`

### Compile the Contract {: #compile-the-contract }

Next, you will need to compile the interface in Remix:

1. Click on the **Compile** tab, second from top

> 2. Compile the interface by clicking on **Compile XcmInterface.sol**

Reviewer 3: ```suggestion
2. Compile the interface by clicking on **Compile XCMInterface.sol**
```

![Compiling XcmInterface.sol](/images/builders/toolkit/ethereum-api/precompiles/xcm-interface/xcm-interface-1.webp)

When the compilation is completed, you will see a green checkmark next to the **Compile** tab.

### Access the Contract {: #access-the-contract }

Instead of deploying the precompile, you will access the interface given the address of the precompiled contract:

1. Click on the **Deploy and Run** tab directly below the **Compile** tab in Remix. Please note that the precompiled contracts are already accessible at their respective addresses. Therefore, there is no deployment step
2. Make sure **Injected Provider - Metamask** is selected in the **ENVIRONMENT** dropdown. Once you select **Injected Provider - Metamask**, you may be prompted by MetaMask to connect your account to Remix if it's not already connected
3. Make sure the correct account is displayed under **ACCOUNT**

> 4. Ensure **XCM - XcmInterface.sol** is selected in the **CONTRACT** dropdown. Given that it is a precompiled contract, there is no deployment step. Instead, you are going to provide the address of the precompile in the **At Address** field

Reviewer 3: ```suggestion
4. Ensure **XCM - XCMInterface.sol** is selected in the **CONTRACT** dropdown. Given that it is a precompiled contract, there is no deployment step. Instead, you are going to provide the address of the precompile in the **At Address** field
```

> 5. Provide the address of the precompile: `{{networks.dancebox.precompiles.xcmInterface}}` and click **At Address**

Reviewer 3: Variable Showing as blank for me now

The **XCM Interface** precompile will appear in the list of **Deployed Contracts**.

### Send Tokens Over to Another EVM-Compatible Appchain {: #transfer-to-evm-chains }

To send tokens over to an account in another EVM-compatible appchain, please follow these steps:

1. Expand the **transferAssetsToPara20** function
2. Enter the appchain ID (paraId)
3. Enter the 20-bytes (Ethereum-like) destination account (beneficiary)
4. Specify the tokens to be transferred. Note that this parameter is an array that contains at least one asset. Each asset is specified by its address and the total amount to transfer

--8<-- 'text/builders/toolkit/ethereum-api/precompiles/xcm-interface/erc-20-note.md'

5. Enter the index of the asset that will be used to pay the fees. This index is zero-based, so the first element is `0`, the second is `1`, and so on 
6. Enter the maximum gas to pay for the transaction. This gas is derived from two parameters, the processing time (`refTime`) and the proof size (`proofSize`). In practice, setting refTime to `uint64::MAX` is equal to *unlimited weight*
7. Click **transact**

> 8. MetaMask will pop up, and you will be prompted to review the transaction details. Click **Confirm** to send the transaction

Reviewer 3: Numbering starts over here for me, this can be avoided by indenting the note in line with the indentation amount of the numbered section. 

<img width="926" alt="Screenshot 2024-08-07 at 10 02 02 AM" src="https://github.com/user-attachments/assets/a3fa080d-baf5-4abf-a58f-eea8d3ba8194">

---
title: Native Cross-Chain Token Transfers

> description: Learn how to use the XCM interface Precompile to transfer tokens from any Tanssi EVM appchain, leveraging their inherent native cross-chain capabilities.

Reviewer: ```suggestion
description: Learn how to use the XCM interface precompile to transfer tokens from any Tanssi EVM appchain, leveraging their inherent native cross-chain capabilities.
```

---

# Native Cross-Chain Token Transfers

## Introduction {: #introduction }

As presented in the [Native Cross-Chain Communication](/learn/framework/xcm/){target=\_blank} article from the Learn section, Tanssi appchains benefit from an inherent capability to communicate and interoperate with any other appchain in the ecosystem. This native cross-chain communication allows safe and fast token transfers leveraging the Cross-Consensus Message format (XCM for short), which facilitates communication between different consensus systems.

The communication protocol enabling token transfers is built on [Substrate](/learn/framework/overview/#substrate-framework){target=\_blank} and runs on a lower level than the EVM, making it harder for EVM developers to access.

> This precompile fills the gap between execution layers, exposing a smart contract that abstracts away the underlying complexities, making the execution of cross-chain token transfers as easy as any other smart contract call.

Reviewer 4: ```suggestion
Nevertheless, EVM Appchains have an XCM precompile that fills the gap between execution layers, exposing a smart contract interface that abstracts away the underlying complexities, making the execution of cross-chain token transfers as easy as any other smart contract call.
```

> This guide will show you how to interact with the [XCM interface](https://github.com/moondance-labs/tanssi/blob/master/test/contracts/solidity/XcmInterface.sol){target=\_blank} precompile to execute cross-chain token transfers.

Reviewer 4: ```suggestion
This guide will show you how to interact with the [XCM interface](https://github.com/moondance-labs/tanssi/blob/master/test/contracts/solidity/XcmInterface.sol){target=\_blank} precompile to execute cross-chain token transfers through the Ethereum API.
```

> The precompile is located at the following address:

Reviewer 4: ```suggestion
The XCM precompile is located at the following address:
```

Reviewer 4: Per what was set in batch and Call Permit, we need to modify the native Token ERC-20

```text
{{networks.dancebox.precompiles.xcmInterface }}
```

--8<-- 'text/builders/toolkit/ethereum-api/precompiles/security-note.md'

## The XCM Solidity Interface {: #the-xcm-solidity-interface }

The [`XCMInterface.sol`](https://github.com/moondance-labs/tanssi/blob/master/test/contracts/solidity/XcmInterface.sol){target=\_blank} interface on Tanssi EVM appchains is a Solidity interface that allows developers to interact with the precompile's functions.

??? code "XcmInterface.sol"

    ```solidity
    --8<-- 'code/builders/toolkit/ethereum-api/precompiles/xcm-interface/XcmInterface.sol'
    ```

The interface includes the necessary data structures along with the following functions:

> ???+ function "**transferAssetsToPara20**(*uint32* paraId, *address* beneficiary, *AssetAddressInfo[] memory* assets, *uint32* feeAssetItem, *Weight memory* weight) — sends assets to another EVM-compatible appchain using the underlying transfer_assets() transaction included in the module called pallet-xcm"

Reviewer: ```suggestion
???+ function "**transferAssetsToPara20**(*uint32* paraId, *address* beneficiary, *AssetAddressInfo[] memory* assets, *uint32* feeAssetItem, *Weight memory* weight) — sends assets to another EVM-compatible appchain using the underlying transfer_assets() transaction included in the pallet-xcm module"
```

>     === "Parameters"

Reviewer 4: I like this a lot! We should do this from now on tbh (cc @eshaben @dawnkelly09 )

Reviewer 4: I think we do something with `transfer_asset()` (cc @dawnkelly09 do you know at the top of your head)?

Reviewer: This is the format we've been trying to move towards! Getting there slowly... only the XCM pages have this right now haha

Although @augustocollado added a nice touch with the example tab 💜

Reviewer 2: > I think we do something with `transfer_asset()` (cc @dawnkelly09 do you know at the top of your head)?

I found this in the comments on the contract:
```
/// @dev Function to send assets via XCM to a 20 byte-like parachain 
/// using transfer_assets() pallet-xcm extrinsic.
```
I'm honestly not quite sure what you're referencing by 'do something' with it though.

Reviewer: The thing that we do with `transfer_asset()`, because it's a function, is wrap it in back ticks. I think that's what he was trying to hint at

        - `paraId` - the destination's appchain ID
        - `beneficiary` - the ECDSA-type account in the destination chain that will receive the tokens
        - `assets` - an array of assets to send
        - `feeAssetItem` - the index of the asset that will be used to pay fees
        - `weight` - the maximum gas to use in the whole operation. Setting uint64::MAX to `refTime` acts in practice as *unlimited weight*

    === "Example"

        - `paraId` - 888
        - `beneficiary` - 0x3f0Aef9Bd799F1291b80376aD57530D353ab0217
        - `assets` - [["0x0000000000000000000000000000000000000800", 1000000000000000000]]
        - `feeAssetItem` - 0
        - `weight` - [9223372036854775807, 9223372036854775807]

> ??? function "**transferAssetsToPara32**(*uint32* paraId, *bytes32* beneficiary, *AssetAddressInfo[] memory* assets, *uint32* feeAssetItem, *Weight memory* weight) — sends assets to a Substrate appchain using the underlying transfer_assets() transaction included in the module called pallet-xcm"

Reviewer 4: Same comment as above, unsure if we do something with `transfer_assets()` check other functions

    === "Parameters"

        - `paraId` - the destination's appchain ID

>         - `beneficiary` - the Substrate's sr25519-type account in the destination chain that will receive the tokens

Reviewer: ```suggestion
        - `beneficiary` - the Substrate's SR25519-type account in the destination chain that will receive the tokens
```

Instead of deploying the precompile, you will access the interface given the address of the precompiled contract:

1. Click on the **Deploy and Run** tab directly below the **Compile** tab in Remix. Please note that the precompiled contracts are already accessible at their respective addresses. Therefore, there is no deployment step
2. Make sure **Injected Provider - Metamask** is selected in the **ENVIRONMENT** dropdown. Once you select **Injected Provider - Metamask**, you may be prompted by MetaMask to connect your account to Remix if it's not already connected
3. Make sure the correct account is displayed under **ACCOUNT**
4. Ensure **XCM - XcmInterface.sol** is selected in the **CONTRACT** dropdown. Given that it is a precompiled contract, there is no deployment step. Instead, you are going to provide the address of the precompile in the **At Address** field
5. Provide the address of the precompile: `{{networks.dancebox.precompiles.xcmInterface}}` and click **At Address**

![Access the address](/images/builders/toolkit/ethereum-api/precompiles/xcm-interface/xcm-interface-2.webp)

The **XCM Interface** precompile will appear in the list of **Deployed Contracts**.

> ### Send Tokens Over to Another EVM-compatible Appchain {: #transfer-to-evm-chains }

Reviewer: ```suggestion
### Send Tokens Over to Another EVM-Compatible Appchain {: #transfer-to-evm-chains }
```

To send tokens over to an account in another EVM-compatible appchain, please follow these steps:

1. Expand the **transferAssetsToPara20** function
2. Enter the appchain ID (paraId)

> 3. Enter the ECDSA destination account (beneficiary)

Reviewer 4: ```suggestion
3. Enter the 20-bytes (Ethereum-like) destination account (beneficiary)
```

4. Specify the tokens to be transferred. Note that this parameter is an array that contains at least one asset. Each asset is specified by its address and the total amount to transfer

> --8<-- 'text/builders/toolkit/ethereum-api/precompiles/xcm-interface/erc-20-note.md'

Reviewer: Here is where we would indent the snippet

Reviewer 2: fixed the snippet and indented it in the markdown where it is used to align as needed

5. Enter the index of the asset that will be used to pay the fees. This index is zero-based, so the first element is `0`, the second is `1`, and so on 

> 6. Enter the maximum gas to pay for the transaction. This gas is derived from two parameters, the processing time (refTime) and the proof size (proofSize). In practice, setting refTime to `uint64::MAX` is equal to *unlimited weight*

Reviewer: ```suggestion
6. Enter the maximum gas to pay for the transaction. This gas is derived from two parameters, the processing time (`refTime`) and the proof size (`proofSize`). In practice, setting refTime to `uint64::MAX` is equal to *unlimited weight*
```

7. Click **transact**
8. MetaMask will pop up, and you will be prompted to review the transaction details. Click **Confirm** to send the transaction

![Confirm Approve Transaction](/images/builders/toolkit/ethereum-api/precompiles/xcm-interface/xcm-interface-3.webp)

After the transaction is confirmed, wait for a few blocks for the transfer to reach the destination chain and reflect the new balance.

### Send Tokens Over to a Substrate Appchain {: #transfer-to-substrate-chains }

To send tokens over to an account in a Substrate appchain, please follow these steps:

1. Expand the **transferAssetsToPara32** function

> 2. Enter the appchain ID (paraId)

Reviewer: ```suggestion
2. Enter the appchain ID (`paraId`)
```

Here you're referencing parameters, so let's wrap all these in back ticks

1. Expand the **transferAssetsToRelay** function
2. Enter the sr25519-type destination account (beneficiary)
3. Specify the tokens to be transferred. Note that this parameter is an array that contains at least one asset. Each asset is specified by its address and the total amount to transfer
    
--8<-- 'text/builders/toolkit/ethereum-api/precompiles/xcm-interface/erc-20-note.md'

4. Enter the index of the asset that will be used to pay the fees. This index is zero-based, so the first element is `0`, the second is `1`, and so on 
5. Enter the maximum gas to pay for the transaction. This gas is derived from two parameters, the processing time (refTime) and the proof size (proofSize). In practice, setting refTime to `uint64::MAX` is equal to *unlimited weight*
6. Click **transact**
7. MetaMask will pop up, and you will be prompted to review the transaction details. Click **Confirm** to send the transaction

![Confirm Approve Transaction](/images/builders/toolkit/ethereum-api/precompiles/xcm-interface/xcm-interface-5.webp)

After the transaction is confirmed, wait for a few blocks for the transfer to reach the destination chain and reflect the new balance.

> ### Send Tokens Over Specifying Locations {: #transfer-locations }

Reviewer 4: This title seems wrong
```suggestion
### Send Tokens Over Specific Locations {: #transfer-locations }
```

> This function is more general than the others, allowing the destination chain, destination account, and assets to be specified using [XCM Multilocations](/learn/framework/xcm/#message-destinations){target=\_blank}.

Reviewer: generic

> To send tokens specifying locations, please follow these steps:

Reviewer: to specific

1. Expand the **transferAssetsLocation** function

> 2. Enter the Multilocation that specifies the destination chain. Note that any chain can be specified, regardless of its configuration or type

Reviewer: no need to capitalize multilocation here

4. Specify the tokens to be transferred. Note that this parameter is an array that contains at least one asset and each asset is specified by its Multilocation and the total amount to transfer
    
--8<-- 'text/builders/toolkit/ethereum-api/precompiles/xcm-interface/erc-20-note.md'

5. Enter the index of the asset that will be used to pay the fees. This index is zero-based, so the first element is `0`, the second is `1`, and so on 
6. Enter the maximum gas to pay for the transaction. This gas is derived from two parameters, the processing time (refTime) and the proof size (proofSize). In practice, setting refTime to `uint64::MAX` is equal to *unlimited weight*
7. Click **transact**
8. MetaMask will pop up, and you will be prompted to review the transaction details. Click **Confirm** to send the transaction

![Confirm Approve Transaction](/images/builders/toolkit/ethereum-api/precompiles/xcm-interface/xcm-interface-6.webp)

After the transaction is confirmed, wait for a few blocks for the transfer to reach the destination chain and reflect the new balance.

> 

Reviewer: ```suggestion
```
